### ISSUE TITLE
Multiple FlowForge Hosted Instances per Application

### ISSUE THREAD

**Issue by joepavitt** [flowforge/flowforge#1689]
### Description

<img width="1261" alt="Screenshot 2023-02-09 at 14 25 37" src="https://user-images.githubusercontent.com/99246719/217839945-94adcfe6-3f3e-4357-98c4-338806745ad4.png">

Two large ticket items we want to deliver to customers are:

- High Availability - https://github.com/flowforge/flowforge/issues/1678
- Staged Deployments & DevOps Pipeline - https://github.com/flowforge/flowforge/issues/1492
    - intermittent delivery in 1.4 - https://github.com/flowforge/flowforge/issues/1580

In both of these issues, to do them _properly_ we would need to support a FlowForge project being able to host multiple instances of Node-RED.

Conceptually, this diagram shows an evolution of the model:

<img width="2564" alt="Screenshot 2023-02-09 at 14 51 10" src="https://user-images.githubusercontent.com/99246719/217846681-8a78a487-0de7-45b7-bbea-57f554d41858.png">

This epic encompasses the discussion and design work to understand the scope of work required to support this, and then to detail the development tasks at hand to make that happen.

Source of this plan comes from https://github.com/flowforge/flowforge/issues/1450#issuecomment-1372564433, but to copy here for easy tracking/discussion:

**Required Changes:**
- Bare minimum requirement here is that each instance would need to have its own Settings page, as per our current "Device" page, for the reason you've raised (env vars)
- Audit Logging needs consideration here for each instance - quick fix would be to just log project.. events where we now have Instances, and with the recent introductions of the new Audit Log formatting, it makes it easy to distinguish between the source of the logs (i.e. which instance they belong to)

**Comments:**
- Under the covers here, we could just transfer everything we have for a "Project", and convert that to the functionality of an Instance, then on the front-end, we package those and arrange them into the new "Projects" concept, which actually keeps the Projects UX as things stand.
- Audit Log filtering becomes more valuable here, and relevant at the "Project" level, as we could filter by "Instance".
- Although this does break our audit logging concepts a bit, as we use project. there.

### UI Rework Task List

```[tasklist]
- [ ] **=== Needed for 1.5 ===**
- [x] flowforge/flowforge#1732
- [ ] flowforge/flowforge#1718
- [ ] flowforge/flowforge#1719
- [ ] https://github.com/flowforge/flowforge/issues/1802
- [ ] https://github.com/flowforge/flowforge/issues/1810
- [ ] https://github.com/flowforge/flowforge/issues/1819
- [ ] https://github.com/flowforge/flowforge-device-agent/issues/58
- [ ] https://github.com/flowforge/flowforge-nr-project-nodes/issues/22
- [ ] https://github.com/flowforge/flowforge-nr-tools-plugin/issues/8
- [ ] https://github.com/flowforge/flowforge/issues/1831
- [ ] https://github.com/flowforge/flowforge/issues/1829
- [ ] https://github.com/flowforge/flowforge/issues/1822
- [ ] https://github.com/flowforge/flowforge/issues/1825
- [ ] https://github.com/flowforge/flowforge/issues/1833
- [ ] https://github.com/flowforge/flowforge/issues/1836
- [ ] https://github.com/flowforge/flowforge/issues/1830
- [ ] https://github.com/flowforge/flowforge/issues/1824
- [ ] https://github.com/flowforge/flowforge/issues/1823
- [ ] **=== Nice to Have 1.5 ===**
- [ ] https://github.com/flowforge/flowforge/issues/1785
- [ ] https://github.com/flowforge/flowforge/issues/1800
- [ ] https://github.com/flowforge/flowforge/issues/1801
```

### 1.6 Release Tasks
```[tasklist]
- [ ] https://github.com/flowforge/flowforge/issues/1735
- [ ] https://github.com/flowforge/flowforge/issues/1734
- [ ] https://github.com/flowforge/flowforge/issues/1737
- [ ] https://github.com/flowforge/flowforge/issues/1851
- [ ] https://github.com/flowforge/flowforge/issues/1831
- [ ] https://github.com/flowforge/flowforge/issues/1849
- [ ] https://github.com/flowforge/flowforge/issues/1859
- [ ] https://github.com/flowforge/flowforge/issues/1914
```

**Comment by knolleary** [comment]
I suggest we can make progress on two fronts in parallel:

**Thread 1**: Make the UI changes to make it look like you can have multiple instances per project, but do not include the 'add instance' button. Build that around the existing platform APIs (ie where each project is its own instance). That moves the UX in the direction we want to go.

**Thread 2**: Work on the underlying model changes needed to support multiple instances per project along with the API changes.

At the appropriate point we can bring those two threads together by updating the UI to use any new APIs - and introduce the 'add instance' button at that point. Alternatively, we fallback to shipping just the Thread 1 work that moves the UX in the right direction, and the 'multiple instance per project' is fully enabled in 1.6.

**Comment by joepavitt** [comment]
@hardillb @knolleary @Pezmc and I currently discussing on a call.

Concern over onboarding for splitting Project > Instance & Application with the onboarding process as it could add more steps to first Node-RED.

General consensus is that the onboarding would wrap the first Application & Instance creation into a single form, very similar to the current "Create Project" form. The only additional field required would be the Application name, e.g.

<img width="1930" alt="Screenshot 2023-02-22 at 10 32 37" src="https://user-images.githubusercontent.com/99246719/220595033-3d547bc6-29e2-4939-951d-a4987a51fb4a.png">

**Comment by Pezmc** [comment]
@joepavitt @knolleary Just an FYI, I'm going to start adding any "other" tasks discovered during the work to the new "Misc, to be sorted, discovered during work" list so they're all in one place. We can create issues/move them down into specific tasks as/when needed.

**Comment by Pezmc** [comment]
Summary of current status at the end of week three:

The UI only work to introduce the instance layer and rename projects to application is in a ready to release state and deployed to staging.

 - #1718 - added the new instance pages (introduced over several PRs)*
 - #1719 - made the changes to the projects pages*
 - Several other PR's (see https://github.com/flowforge/flowforge/pull/1792) made general renaming, and cleanup.

*<sup>(these issues all have several sub-issues)</sup>

Two smaller UI only tasks (which _could_ be follow ups) are open:

 - Moving settings between Instances and Applications, currently there are no settings at the application level: https://github.com/flowforge/flowforge/issues/1734

 - Billing UX at the application level: https://github.com/flowforge/flowforge/issues/1737

**Comment by Pezmc** [comment]
The task list above now contains extra tasks raised as part of the UX review, additionally I've split the list into "for 1.5", nice to have in 1.5, and follow ups for 1.6 to give an easier overview of what we should aim to complete this week.

**Comment by joepavitt** [comment]
Can now close this, functional on staging, outstanding items captured in #1972 - none of which are blocking or significant bugs, all feature improvements
